pageserver/src/tenant_threads.rs

```diff
@@ -1,78 +1,229 @@
+        .get()
+        .context("Failed to get START_GC_LOOP")?
+        .blocking_send(tenantid)
+        .map_err(|e| anyhow::anyhow!(e))
+        .context("Failed to send to START_GC_LOOP channel")?;
+    Ok(())
+}
+
+/// Spawn a task that will periodically schedule compaction until
+/// the tenant becomes inactive. This should be called on tenant
+/// activation.
+pub fn start_compaction_loop(tenantid: ZTenantId) -> anyhow::Result<()> {
+    START_COMPACTION_LOOP
+        .get()
+        .context("failed to get START_COMPACTION_LOOP")?
+        .blocking_send(tenantid)
+        .map_err(|e| anyhow::anyhow!(e))
+        .context("failed to send to START_COMPACTION_LOOP")?;
+    Ok(())
+}
+
+/// Spawn the TenantTaskManager
+/// This needs to be called before start_gc_loop or start_compaction_loop
+pub fn init_tenant_task_pool() -> anyhow::Result<()> {
+    let runtime = tokio::runtime::Builder::new_multi_thread()
+        .thread_name("tenant-task-worker")
+        .worker_threads(40) // Way more than necessary
+        .max_blocking_threads(100) // Way more than necessary
+        .enable_all()
+        .build()?;
+
+    let (gc_send, mut gc_recv) = mpsc::channel::<ZTenantId>(100);
+    START_GC_LOOP
+        .set(gc_send)
+        .expect("Failed to set START_GC_LOOP");
+
+    let (compaction_send, mut compaction_recv) = mpsc::channel::<ZTenantId>(100);
+    START_COMPACTION_LOOP
+        .set(compaction_send)
+        .expect("Failed to set START_COMPACTION_LOOP");
+
+    // TODO this is getting repetitive
+    let mut gc_loops = HashMap::<ZTenantId, watch::Sender<()>>::new();
+    let mut compaction_loops = HashMap::<ZTenantId, watch::Sender<()>>::new();
+
+    thread_mgr::spawn(
+        ThreadKind::TenantTaskManager,
+        None,
+        None,
+        "Tenant task manager main thread",
+        true,
+        move || {
+            runtime.block_on(async move {
+                loop {
+                    tokio::select! {
+                        _ = thread_mgr::shutdown_watcher() => {
+                            // TODO do this from thread_mgr? Extend it to work with tasks?
+                            for (_, cancel) in gc_loops.drain() {
+                                cancel.send(()).ok();
+                            }
+                            for (_, cancel) in compaction_loops.drain() {
+                                cancel.send(()).ok();
+                            }
+                            // TODO wait for tasks to die?
+                            break;
+                        },
+                        tenantid = gc_recv.recv() => {
+                            let tenantid = tenantid.expect("Gc task channel closed unexpectedly");
+
+                            // Spawn new task, request cancellation of the old one if exists
+                            let (cancel_send, cancel_recv) = watch::channel(());
+                            let _handle = tokio::spawn(gc_loop(tenantid, cancel_recv))
```

Comment:
`gc_loop` does a bunch of `anyhow::bail!` and might even panic when calling `tenant_mgr::get_tenant_state` for logging.

The only way to know about this is to do `_handle.await` (or call other methods on the `_handle`) and to process the end result, otherwise we will never know how did the task end: we can add channels inside and know that the task had ended by the fact that the channel counterpart got dropped, but the end result (due to panic possibilities) has to be always checked on the handle.

This brings two implications:
* if anything fails in the GC/compaction thread now, we would never see the end error logged
* we don't restart such task before another set of `set_tenant_state` calls, meaning we may silently loose both compaction and gc tasks if very unlucky


Same applies to compaction (quite a routine comment in this PR related to tenant task management, I wonder if the logic is worth deduplicating already at this point)

Reply:
Added an await with `FuturesUnordered`. I tested with a manual failpoint to see that the panic gets logged